### ISSUE TITLE
curl_multi_perform not returning for 4+ minutes

### ISSUE THREAD

**Issue by stolcman** [curl/curl#12188]
### I did this

- connect to hotspot
- disable WAN (so we have network connectivity but no internet)
- issue a POST on an HTTP2 stream

When this happens, we get stuck in a while loop in h2_progress_egress function in lib/http2.c.  I put some extra logs:

Enter the loop:

```
2023-10-23 16:58:12.870 [   b5] I CURLINFO_TEXT* ***** readwrite_upload - before Curl_write *****
2023-10-23 16:58:12.871 [   b5] I CURLINFO_TEXT* ***** Curl_write - before conn->send *****
2023-10-23 16:58:12.871 [   b5] I CURLINFO_TEXT* ***** cf_h2_send - before h2_progress_egress *****
2023-10-23 16:58:12.871 [   b5] I CURLINFO_TEXT* ***** h2_progress_egress - begin *****
2023-10-23 16:58:12.871 [   b5] I CURLINFO_TEXT* ***** h2_progress_egress - before while loop *****
```
nghttp2_session_send returns 0 for over 4 minutes:
```
2023-10-23 16:58:12.871 [   b5] I CURLINFO_TEXT* ***** h2_progress_egress - nghttp2_session_send returned 0 *****
2023-10-23 16:58:12.871 [   b5] I CURLINFO_TEXT* ***** h2_progress_egress - nghttp2_session_send returned 0 *****
2023-10-23 16:58:12.871 [   b5] I CURLINFO_TEXT* ***** h2_progress_egress - nghttp2_session_send returned 0 *****
...
2023-10-23 17:02:27.129 [   b5] I CURLINFO_TEXT* ***** h2_progress_egress - nghttp2_session_send returned 0 *****
2023-10-23 17:02:27.130 [   b5] I CURLINFO_TEXT* ***** h2_progress_egress - nghttp2_session_send returned 0 *****
2023-10-23 17:02:27.130 [   b5] I CURLINFO_TEXT* ***** h2_progress_egress - nghttp2_session_send returned 0 *****
2023-10-23 17:02:27.130 [   b5] I CURLINFO_TEXT* ***** h2_progress_egress - nghttp2_session_send returned 0 *****
2023-10-23 17:02:27.130 [   b5] I CURLINFO_TEXT* ***** h2_progress_egress - nghttp2_session_send returned 0 *****
2023-10-23 17:02:27.130 [   b5] I CURLINFO_TEXT* ***** h2_progress_egress - nghttp2_session_send returned 0 *****
```
... until it finally errors out
```
2023-10-23 17:02:27.130 [   b5] I CURLINFO_TEXT* Send failure: Connection timed out
2023-10-23 17:02:27.136 [   b5] I CURLINFO_TEXT* BoringSSL SSL_write: Connection timed out, errno 110
2023-10-23 17:02:27.136 [   b5] I CURLINFO_TEXT* Failed sending HTTP2 data

2023-10-23 17:02:27.136 [   b5] I CURLINFO_TEXT* ***** h2_progress_egress - nghttp2_session_send returned -902 *****
2023-10-23 17:02:27.136 [   b5] I CURLINFO_TEXT* ***** h2_progress_egress - after while loop *****
2023-10-23 17:02:27.155 [   b5] I CURLINFO_TEXT* ***** cf_h2_send - after h2_progress_egress *****
2023-10-23 17:02:27.155 [   b5] I CURLINFO_TEXT* ***** Curl_write - after conn->send *****
2023-10-23 17:02:27.155 [   b5] I CURLINFO_TEXT* ***** readwrite_upload - after Curl_write *****
2023-10-23 17:02:27.156 [   b5] I CURLINFO_TEXT* Send failure: Broken pipe
2023-10-23 17:02:27.156 [   b5] I CURLINFO_TEXT* BoringSSL SSL_write: Broken pipe, errno 32
2023-10-23 17:02:27.156 [   b5] I CURLINFO_TEXT* Failed sending HTTP2 data
```
Do you think we should have some safety code preventing the loop from getting stuck for so long, and consequently preventing curl_multi_perform from returning?  Or is it okay to rely on the lower layers to error out sooner?

### I expected the following

I expected curl_multi_perform to return in a reasonable amount of time

### curl/libcurl version

libcurl version=8.1.2
ssl_version=BoringSSL
libz_version=1.2.8

### operating system

# uname -a
Linux localhost 4.4.22+ #89 SMP PREEMPT Mon Oct 23 17:35:34 UTC 2023 armv7l

**Comment by stolcman** [comment]
The loop has an extra condition in the current version.  Maybe this is already fixed?
```
  ctx->nw_out_blocked = 0;
  while(!rv && !ctx->nw_out_blocked && nghttp2_session_want_write(ctx->h2))
    rv = nghttp2_session_send(ctx->h2);
```
In my version, it's just:
```
 while(!rv && nghttp2_session_want_write(ctx->h2))
    rv = nghttp2_session_send(ctx->h2);
```

**Comment by stolcman** [comment]
I can confirm the issue is fixed in 8.4.0
